doc/IPset.md

```diff
@@ -2,218 +2,709 @@ DPVS IPset
+
+Finally, let's make some tests.
+
+```bash
+# ./bin/dpip ipset -v test foo 10.132.1.2,udp:10240
+10.132.1.2,udp:10240 is in set foo
+# ./bin/dpip ipset -v test foo 10.100.100.100,udp:10240
+10.100.100.100,udp:10240 is NOT in set foo
+# ./bin/dpip ipset test foo 192.168.88.0,tcp:8080
+true
+# ./bin/dpip ipset test foo 192.168.88.22,tcp:8082
+true
+# ./bin/dpip ipset test foo 192.168.88.222,tcp:8082
+false
+```
+The last test returned "false" because of the nomatch entry `192.168.88.200-192.168.88.255,tcp:8082 nomatch`. If we add one more specific entry `192.168.88.222/32,tcp:8082` and test again, the result would turn out to be "true".
+
+```bash
+# ./bin/dpip ipset add foo 192.168.88.222,tcp:8082 comment "you are an exception"
+# ./bin/dpip ipset list -v
+Name: foo
+Type: hash:net,port
+Header: family inet  hashsize 256  maxelem 1000  comment
+Size in memory: 5312
+References: 0
+Number of entries: 8
+Members:
+192.168.88.222/32,tcp:8082  comment "you are an exception"
+192.168.88.200/29,tcp:8082 nomatch  comment "bad guys"
+192.168.88.208/28,tcp:8082 nomatch  comment "bad guys"
+192.168.88.224/27,tcp:8082 nomatch  comment "bad guys"
+192.168.88.0/24,tcp:8080  
+192.168.88.0/24,tcp:8081  
+192.168.88.0/24,tcp:8082  
+10.132.0.0/16,udp:10240  
+
+# ./bin/dpip ipset test foo 192.168.88.222,tcp:8082
+true
+```
+
+<a id='types'/>
+
+# Set Types
+
+> For more exmaple of ipset types, refer to [ipset test script](../test/ipset/dpip.sh) and the [test results](../test/ipset/dpip.log).
+
+<a id='bitmap_ip'/>
+
+#### bitmap:ip
+
+The bitmap:ip set type uses a memory range to store IPv4 addresses. IPv4 range is supported when add/delete entries, but the ipset parses the range and finds all specific IPv4 addresses in the range, then save them in the bitmap storage. Therefore, we can add an IPv4 range to bitmap:ip, but delete a subrange IPv4 address from the ipset. IPv6 is not supported by the type.
+
+```bash
+# ./bin/dpip ipset create foo bitmap:ip range 192.168.0.0/16
+# ./bin/dpip ipset add foo 192.168.1.0/29
+# ./bin/dpip ipset list foo
+Name: foo
+Type: bitmap:ip
+Header: range 192.168.0.0/16  
+Size in memory: 8240
+References: 0
+Number of entries: 8
+Members:
+192.168.1.0  
+192.168.1.1  
+192.168.1.2  
+192.168.1.3  
+192.168.1.4  
+192.168.1.5  
+192.168.1.6  
+192.168.1.7  
+# ./bin/dpip ipset del foo 192.168.1.4-192.168.1.6
+# ./bin/dpip ipset list foo
+Name: foo
+Type: bitmap:ip
+Header: range 192.168.0.0/16  
+Size in memory: 8240
+References: 0
+Number of entries: 5
+Members:
+192.168.1.0  
+192.168.1.1  
+192.168.1.2  
+192.168.1.3  
+192.168.1.7 
+```
+
+<a id='bitmap_ip_mac'/>
+
+#### bitmap:ip,mac
+
+The bitmap:ip,mac set type uses a memory range to store IPv4 and a MAC address pairs. If MAC address is not specified with a entry, then MAC address is ignored when match against this entry, i.e., only IPv4 address is considered in this case.
```

Comment:
If MAC address is not specified in the pair, it is ignored when matching against this entry,